### PR TITLE
Ansible syntax check INFRA-4266

### PR DESCRIPTION
**Pull request by b00men**
_groovy classes for using of Ansible Tower's API and method for ansible syntax checking_

This PR fixes:
- [INFRA-4266](https://jira.anchorfree.net/browse/INFRA-4266) jenkins-pipeline-shared var that uses ansible-tower to check ansible syntax

Examples
---

- Good results, repo is crypted: [pipeline](https://jenkins.anchorfree.net/blue/organizations/jenkins/anchorfree%2Fansible-test-ci/detail/master/173/pipeline/), [checked repo](https://github.com/AnchorFree/ansible-test-ci/tree/1b34782a8f0330f1c016e6ea92c2fb6d1052120a)
- Good results, repo isn't crypted: [pipeline](https://jenkins.anchorfree.net/blue/organizations/jenkins/anchorfree%2Fansible-test-ci/detail/INFRA-4266-without-crypt/1/pipeline/), [checked repo](https://github.com/AnchorFree/ansible-test-ci/tree/a618bbe441132e09ef4c44e3c0b42504c7aef517)
- Dry run failed due error in playbook (undefined variable): [pipeline](https://jenkins.anchorfree.net/blue/organizations/jenkins/anchorfree%2Fansible-test-ci/detail/INFRA-4266-bad-playbook/2/pipeline/), [checked repo](https://github.com/AnchorFree/ansible-test-ci/tree/c2279e853ef4e0263efbfa87df8cda8086e72e62)
- Dry run failed due inaccessible host: [pipeline](https://jenkins.anchorfree.net/blue/organizations/jenkins/anchorfree%2Fansible-test-ci/detail/master/172/pipeline/), [checked repo](https://github.com/AnchorFree/ansible-test-ci/tree/1b34782a8f0330f1c016e6ea92c2fb6d1052120a)

### CODE REVIEW

File: vars/ansibleCheckSyntax.groovy [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+    if (repo.startsWith('ansible-')) {
+        repo = repo.substring(8)
+    }
+    def name = "${repo}(${sha.substring(0,7)})" // general name of awx's objects
+    def unlock_playbook = "unlock.yml"
+    def unlock_vars = "git_crypt_key_path: /tmp/hssunified-ubuntu.key"
+    def awx_host = "awx.afdevops.com"
+    def awx_cred_id = 2 // for github pulling
+    def awx_ssh_id = 6 // for servers' access
+    def awx_org_id = 2
```
Comment (ahammond) [inline vars/ansibleCheckSyntax.groovy]: These magic number need explaining please.

File: vars/ansibleCheckSyntax.groovy [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+    def branch = tokens[tokens.size() - 1].replaceAll('%2F','-')
+    def sha = sh(returnStdout: true, script: 'git rev-parse HEAD').trim()
+    def scm_path = "${org}/${repo}"
+    // if name starts with ansible-, remove the prefix.
+    if (repo.startsWith('ansible-')) {
+        repo = repo.substring(8)
+    }
+    def name = "${repo}(${sha.substring(0,7)})" // general name of awx's objects
+    def unlock_playbook = "unlock.yml"
+    def unlock_vars = "git_crypt_key_path: /tmp/hssunified-ubuntu.key"
```
Comment (ahammond) [inline vars/ansibleCheckSyntax.groovy]: This breaks Jenkins' secret protection. It's fine for a prototype, but we'll need to refactor to use a with credentials like https://github.com/AnchorFree/jenkins-pipeline-shared/blob/master/vars/dockerBuildTagPush.groovy#L31

File: README.md [hunk 3]
```diff
@@ -56,6 +56,46 @@ The above will protect only the branches with the name `protected` and `other_pr
...
 
+### Ansible "Dry Run"
+
+For automatic syntax checking of ansible playbooks you can use method `ansibleCheckSyntax` with next arguments:
+- List of playbooks, which should be tested;
+- Inventory file which will be used for dry run;
+
+There are several important things:
+- Before list you mast define type `(String[])` (see an example below) or use one string with `split()` method (see an example from [**Protected Docker Tags**](#protected-docker-tags));
+- Host from inventory for dry run should be accessible;
```
Comment (ahammond) [inline README.md]: Accessible by what user? What ssh public key?

File: README.md [hunk 4]
```diff
@@ -56,6 +56,46 @@ The above will protect only the branches with the name `protected` and `other_pr
...
+- hosts: all
+  gather_facts: False
+  pre_tasks:
+    - name: Unlock repository
+      local_action: shell git-crypt unlock {{ git_crypt_key_path }}
+    - name: Unlock repository common
+      local_action: shell cd roles/common/ && git-crypt unlock {{ git_crypt_key_path }}
+```
+
+Example of `Jenkins` file. `inventory-dry` is a inventory for syntax check, `some_playbook.yml` and `another_playbook.yml` is playbooks which will be checked.
```
[the remark this answers is not in the document]
Comment (b00men) [reply README.md]: Fixed in da92f8a4cf14e124c28ba9f94f0c055fd893639f

File: src/com/anchorfree/AnsibleTowerApi.groovy [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+		try {
+			subj = waitStatus()
+			if (subj.status != "successful") {
+				awx.failed=true
+				awx.error_messages.add("${name} (${type}) status is ${subj.status}")
+			}			
+		}
+		catch(java.lang.NullPointerException e) {
+			awx.failed=true
+			awx.error_messages.add("Can't check status. Probably ${name}(${type}) didn't created: "+e.getMessage())
```
[the remark this answers is not in the document]
Comment (b00men) [reply src/com/anchorfree/AnsibleTowerApi.groovy]: Fixed in 327aedc6996949e6be9bc48b7b89cf6f49584d2d

File: src/com/anchorfree/JenkinsHttpClient.groovy [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+                .basicAuthentication(user, password)
+                .protocol("https")
+                .host(host)
+                .path(path)
+                .port(443)
+                .header("User-Agent", userAgent)
+                .send()
+        return resp
+    }
+}
```
[the remark this answers is not in the document]
Comment (b00men) [reply src/com/anchorfree/JenkinsHttpClient.groovy]: Fixed

File: vars/ansibleCheckSyntax.groovy (hunk 2 above)
Comment (b00men) [reply vars/ansibleCheckSyntax.groovy]: Fixed in 7528b3cac74b304b9f509be610ff74d9850758b7

File: vars/ansibleCheckSyntax.groovy (hunk 1 above)
Comment (b00men) [reply vars/ansibleCheckSyntax.groovy]: Fixed in 7528b3cac74b304b9f509be610ff74d9850758b7

File: README.md (hunk 3 above)
Comment (b00men) [reply README.md]: Fixed in 7528b3cac74b304b9f509be610ff74d9850758b7 and described in da92f8a4cf14e124c28ba9f94f0c055fd893639f. 
Actually we awaiting merge of [ARC#200](https://github.com/AnchorFree/ansible-roles-common/pull/200) for real using of this feature.

File: src/com/anchorfree/AnsibleTowerApiJobTemplate.groovy [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+		type = "job_templates"
+	}
+
+	// We can't call awx' REST API from constructors due https://issues.jenkins-ci.org/browse/JENKINS-26313
+	// So we had to create make() method
+	def make() {
+		try {
+		    def messageBody = ['name': name,
+		        'description': "Created by jenkins",
+				'job_type': job_type,
```
[the remark this answers is not in the document]
Comment (b00men) [reply src/com/anchorfree/AnsibleTowerApiJobTemplate.groovy]: Fixed in 28d30193c53b4ac4e1a170250fda16e0203cbd68

File: src/com/anchorfree/AnsibleTowerApiInventory.groovy [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+		}
+		catch(java.lang.NullPointerException e) {
+			awx.failed=true
+			awx.error_messages.add("Can't create inventory source ${inventory_source_name}. Probably ${project.name}(${project.type}) didn't created: \n"+e.getMessage())
+			return null
+		}
+
+	}
+
+	/** We can leave inventory_source because it will be removed by awx as soon as related inventory will be removed
```
[the remark this answers is not in the document]
Comment (b00men) [reply src/com/anchorfree/AnsibleTowerApiInventory.groovy]: Fixed in 1c86e1fc07cf494d71684386f80318477cbfff23

File: src/com/anchorfree/AnsibleTowerApi.groovy [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+
+	def update(String path = "api/v2/${type}/${subj.id}/" ) {
+	    def response = new JenkinsHttpClient().get(awx.host, path, awx.user, awx.password)
+	    if (checkResponse(response, "Can't get status of ${path}") != true ) { return null } 
+		return new groovy.json.JsonSlurper().parseText(response.bodyText())
+	}
+
+	def waitStatus(obj = subj, String path = "api/v2/${type}/${subj.id}/" ) {
+		while( obj.status =~ /^(new)|(waiting)|(pending)|(running)$/) {
+			sleep(5000)
```
[the remark this answers is not in the document]
Comment (b00men) [reply src/com/anchorfree/AnsibleTowerApi.groovy]: AWX provide such possibility but I going to implement it in our side in next time as minor feature to give acceleration for main project.
I've created corresponding [ticket](https://jira.anchorfree.net/browse/INFRA-4563) for fixation of necessity of this changes.

**Pull request merged**